master/buildbot/worker/gce.py

```diff
@@ -0,0 +1,228 @@
+
+        instance_stop = None
+        if instance_state['status'] not in ('STOPPED', 'TERMINATED'):
+            log.info("gce: {0} is running, requesting stop".format(
+                self.instance))
+            instance_stop = self._gce.instanceStop()
+
+        if not self.resetDisk:
+            metadata['BUILDBOT_CLEAN'] = '1'
+
+        boot_disk_create = None
+        if 'BUILDBOT_CLEAN' not in metadata:
+            log.info("gce: {0} has not been reset on stop, will do it now".format(
+                self.instance))
+            boot_disk_name = self.getNewDiskName(updated_metadata)
+            boot_disk_create = self.createBootDisk(boot_disk_name)
+
+        metadata_set = self.setMetadata(fingerprint, updated_metadata)
+
+        if instance_stop is not None:
+            log.info("gce: waiting for {0} to be stopped".format(self.instance))
+            yield self._gce.processRequest(instance_stop)
+            yield self._gce.waitInstanceState('TERMINATED')
+
+        if boot_disk_create is not None:
+            current_disk_name = self.getCurrentDiskName(instance_state['disks'])
+            if current_disk_name:
+                log.info("gce: detaching {0} from {1}".format(
+                    current_disk_name, self.instance))
+                yield self._gce.processAsyncRequest(self._gce.detachDisk(current_disk_name))
+                log.info("gce: deleting {0}".format(current_disk_name))
+                yield self._gce.deleteDisk(current_disk_name)
+
+            log.info("gce: waiting for fresh disk {0} to be created for {1}".format(
+                boot_disk_name, self.instance))
+            yield self._gce.processAsyncRequest(boot_disk_create)
+
+            log.info("gce: attaching new disk {0} to {1}".format(
+                boot_disk_name, self.instance))
+            yield self._gce.processAsyncRequest(
+                self._gce.attachDisk(boot_disk_name, index=0, boot=True))
+
+        log.info("gce: starting {0}".format(self.instance))
+        yield self._gce.processAsyncRequest(metadata_set)
+        yield self._gce.processRequest(
+            self._gce.instanceStart())
+        yield self._gce.waitInstanceState('RUNNING')
+        return defer.returnValue(True)
+
+    @defer.inlineCallbacks
+    def stop_instance(self, fast=False):
+        if not self.stopInstanceOnStop:
+            log.info("gce: not stopping {0}: stopInstanceOnStop == False".format(
+                self.instance))
+            return defer.returnValue(None)
+
+        state = yield self._gce.processRequest(self._gce.getInstanceState())
+        if state['status'] not in ('TERMINATED'):
+            log.info("gce: stopping {0}".format(self.instance))
+            yield self._gce.processRequest(self._gce.instanceStop())
+            # We don't use waitOperationEnd here as I'm having ~55s between
+            # the instance state switching to TERMINATED and the operation
+            # reporting to be DONE
```

Comment:
Is it possible to start the instance during these 55 seconds normally? Also, perhaps worth rewording the comment to something like `... as in certain cases as much as 55 seconds pass between ...` because it's not clear who "I" refers to and ideally it should not matter :-).